stablehlo/dialect/Base.cpp

```diff
@@ -261,67 +267,114 @@ FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
-  for (auto inputType : inputTypes)
-    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
-      rankedTypes.push_back(rankedType);
-  if (rankedTypes.empty()) {
-    inferredReturnTypes.push_back(inputTypes[0]);
-    return success();
+// Inference rules for conditional branches (lhs/rhs are commutative):
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               ?
+//  c2:  X              ?, B            ?, max(X, B)
+//  c3:  ?              ?               ?
+//  c4:  ?              ?, B            ?
+//  c5:  ?, B           ?, C            ?, max(B, C)
+FailureOr<std::pair<int64_t, int64_t>> inferLeastSpecificDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound) {
+  bool isLeftStaticDim = !isDynamicDimSize(leftSize);
+  bool isRightStaticDim = !isDynamicDimSize(rightSize);
+  bool isLeftStaticBound = !isDynamicDimSize(leftBound);
+  bool isRightStaticBound = !isDynamicDimSize(rightBound);
+  int64_t inferredSize = ShapedType::kDynamic;
+  int64_t inferredBound = ShapedType::kDynamic;
+
+  if (isLeftStaticDim || isRightStaticDim) {
+    if (isLeftStaticDim && isRightStaticDim) {
+      if (leftSize != rightSize)
+        return emitOptionalError(location, "Mismatched dimension sizes ",
+                                 leftSize, " and ", rightSize, " in dimension ",
+                                 dim);
+      inferredSize = leftSize;
+    } else if (isLeftStaticBound || isRightStaticBound) {
+      inferredBound = isLeftStaticDim ? std::max(leftSize, rightBound)
+                                      : std::max(rightSize, leftBound);
+    }
+  } else if (isLeftStaticBound && isRightStaticBound) {
+    inferredBound = std::max(leftBound, rightBound);
   }
+  return std::make_pair(inferredSize, inferredBound);
+}
 
+FailureOr<TensorType> inferTypeWithCustomFunc(
+    Optional<Location> location, SmallVector<RankedTensorType> rankedTypes,
+    std::function<FailureOr<std::pair<int64_t, int64_t>>(
+        Optional<Location>, int64_t, int64_t, int64_t, int64_t, int64_t)>
+        inferDimAndBoundFunc) {
   auto rank = rankedTypes[0].getRank();
-  SmallVector<int64_t> inferredSizes(rank, ShapedType::kDynamic);
+  SmallVector<int64_t> inferredSizes;
   SmallVector<int64_t> inferredBounds(rank, ShapedType::kDynamic);
   bool anyInputHaveBounds = false;
 
-  for (const auto& it : llvm::enumerate(rankedTypes)) {
-    RankedTensorType rankedType = it.value();
+  for (unsigned i = 0; i < rankedTypes.size(); ++i) {
+    RankedTensorType rankedType = rankedTypes[i];
     ArrayRef<int64_t> bounds = encodingToBounds(rankedType.getEncoding());
     if (!bounds.empty()) anyInputHaveBounds = true;
 
+    if (i == 0) {
+      inferredSizes = to_vector(rankedType.getShape());
+      if (!bounds.empty()) inferredBounds = to_vector(bounds);
+      continue;
+    }
```

Comment:
How about moving the `i == 0` part of the loop above where `inferredSizes` and `inferredBounds` are declared and then start the loop with `i = 1`? This way the variables will be consistently initialized where they are declared, and the loop won't have a special case.

Reply:
Makes sense! Updated.